misc/mdns/src/dns.rs

```diff
@@ -211,6 +203,37 @@ fn append_u16(out: &mut Vec<u8>, value: u16) {
     out.push((value & 0xff) as u8);
 }
 
+/// If a peer ID is longer than 63 characters, split it into segments to
+/// be compatible with RFC 1035.
+fn segment_peer_id(peer_id: &String) -> String {
+    // This will only perform one allocation except in extreme circumstances.
```

Comment:
I think you can replace the function's body with something like this:

```suggestion
    String::from_utf8(peer_id.as_bytes().chunks(63).join(&b'.')).expect("we copy from a UTF8 string; qed")
```

Reply:
The problem is that `.chunks` can't `.join` (in 1.83.0) ... I went through several iterations to see if I could make that work elegantly, and ended up with the `for ... in` loop for clarity and minimal number of allocations.

Reply:
I looked further into this as I was curious, I guess not because it really matters in this case. I am sorry if this is just useless noise to the discussion.

Under the assumption that the string only contains ASCII characters, why not do:

```Rust
    let b = peer_id
        .as_bytes()
        .chunks(MAX_LABEL_LENGTH)
        .collect::<Vec<_>>()
        .join(&b'.');
    String::from_utf8(b).expect("we copy from a UTF8 string; qed")
```

Using a for-loop and manually preallocating the right amount of memory seems like an early optimization to me. I have benchmarked this [here](https://github.com/mxinden/segment/blob/master/benches/lib.rs). Using iterators with `join` allows one to not copy the characters one by one as in the for loop, but instead in chunks (See https://doc.rust-lang.org/src/alloc/slice.rs.html#664).

Running the above linked benchmarks in criterion shows that the above would be twice as fast (on my machine) as the for loop. In addition I find it easier to comprehend.

@peat @tomaka  what do you think?